### PR TITLE
Add sequential gray encoding for enum

### LINKED ISSUE #939: Add gray encoding to Enum encodings
Useful for FSM

### PR DESCRIPTION
**Pull request by andreasWallner**
Fixes #939

RTD addition upcoming^^

### CODE REVIEW

File: core/src/main/scala/spinal/core/Enum.scala [hunk 1]
```diff
@@ -349,6 +349,29 @@ object binaryOneHot extends SpinalEnumEncoding{
...
+/**
+ * Gray encoding (sequentially assigned)
+ * @example{{{ 000, 001, 011, 010, ... }}}
+ * @note If used in FSM is is not ensured that only gray encoding preserving
+ *       transitions are done. If that is needed e.g. for CDC reasons, the
+ *       transitions must be checked manually.
+ */
+object graySequential extends SpinalEnumEncoding {
+  private def gray2binary(gray: BigInt) = {
+    assert(gray.bitLength <= 16, "conversion currently only implemented for <= 16 bit")
```
Comment (Readon) [inline core/src/main/scala/spinal/core/Enum.scala:361]: Maybe it's better to keep 16 a variable but constant.

File: tester/src/test/scala/spinal/tester/scalatest/SpinalSimVerilatorIoTest.scala [hunk 2]
```diff
@@ -201,10 +202,12 @@ class SpinalSimVerilatorIoTest extends AnyFunSuite {
           newEnumTest(dut.nativeEncoding)
           newEnumTest(dut.binaryOneHotEncoding)
           newEnumTest(dut.binarySequentialEncoding)
+          newEnumTest(dut.graySequentialEncoding)
         }
       }
     }
 
+    println(prefix)
```
Comment (Readon) [inline tester/src/test/scala/spinal/tester/scalatest/SpinalSimVerilatorIoTest.scala:210]: Please be careful on the print-out statements.

File: core/src/main/scala/spinal/core/Enum.scala [hunk 3]
```diff
@@ -349,6 +349,29 @@ object binaryOneHot extends SpinalEnumEncoding{
   setName("oh")
 }
 
+/**
+ * Gray encoding (sequentially assigned)
+ * @example{{{ 000, 001, 011, 010, ... }}}
+ * @note If used in FSM is is not ensured that only gray encoding preserving
```
Comment (numero-744) [inline core/src/main/scala/spinal/core/Enum.scala:355]:
```suggestion
 * @note If used in FSM it is not ensured that only gray encoding preserving
```

File: core/src/main/scala/spinal/core/Enum.scala (hunk 1 above)
Comment (numero-744) [reply core/src/main/scala/spinal/core/Enum.scala:361]: Or define a generic function?

```scala

object Test extends App {
  private def gray2binary(gray: BigInt) = {
    @tailrec def aux(g: BigInt, b: BigInt): BigInt = {
      if (g > 0) {
        val g1 = g >> 1
        aux(g1, b ^ g1)
      } else {
        b
      }
    }
    aux(gray, gray)
  }
  private def binary2gray(bin: BigInt) = bin ^ (bin >> 1)

  // helpers for tests
  def assertions(i: BigInt): Unit = {
    assert(gray2binary(binary2gray(i)) == i)
    assert(binary2gray(gray2binary(i)) == i)

    if (i > 0)
      assert(isPow2(binary2gray(i) ^ binary2gray(i - 1)))
  }
  // basic test
  for (n <- 0 until 1000) {
    val i = BigInt(n)

    def toString(x: BigInt): String = {
      val digits = x.toString(2)
      val n = 4
      val l = digits.length
      if (l <= n) {
        "0" * (n - l) + digits
      } else digits
    }
    if (i < 16)
      println((toString(i), toString(binary2gray(i))))

    assertions(i)
  }

  // random test
  for (n <- 0 until 1000){
    def r: BigInt = {
      val x = scala.util.Random.nextLong()
      if (x < 0) r
      else BigInt(x)
    }
    assertions(r)
  }
}
```

Comment (numero-744) [reply core/src/main/scala/spinal/core/Enum.scala:361]: Also, maybe these two helper functions should be available, in `misc` for instance?

File: core/src/main/scala/spinal/core/Enum.scala [hunk 4]
```diff
@@ -349,6 +349,29 @@ object binaryOneHot extends SpinalEnumEncoding{
...
+    tmp ^= tmp >> 4
+    tmp ^= tmp >> 2
+    tmp ^= tmp >> 1
+    tmp
+  }
+  private def binary2gray(bin: BigInt) = bin ^ (bin >> 1)
+
+  override def getWidth(e: SpinalEnum) = log2Up(e.elements.length)
+  override def getValue[T <: SpinalEnum](element: SpinalEnumElement[T]): BigInt = binary2gray(element.position)
+  override def getElement[T <: SpinalEnum](value: BigInt, enums: T): SpinalEnumElement[T] = enums.elements(binary2gray(value).toInt)
```
Comment (numero-744) [inline core/src/main/scala/spinal/core/Enum.scala:372]: I'm wondering, shouldn't it be `gray2binary` here?

Comment (Readon) [reply core/src/main/scala/spinal/core/Enum.scala:372]: There is a "toGray" and "fromGray" [here](https://github.com/SpinalHDL/SpinalHDL/blob/dda945dfafa3e479c680a96385a94bd8e54f929a/lib/src/main/scala/spinal/lib/Utils.scala#L369).
Although, it's a spinal one, but not the Scala function, it's still a good reference.

File: core/src/main/scala/spinal/core/Enum.scala (hunk 1 above)
Comment (andreasWallner) [reply core/src/main/scala/spinal/core/Enum.scala:361]: are we really discussing about the finer detail of having more than 2^16 enum values / fsm states? xD

File: tester/src/test/scala/spinal/tester/scalatest/SpinalSimVerilatorIoTest.scala (hunk 2 above)
Comment (andreasWallner) [reply tester/src/test/scala/spinal/tester/scalatest/SpinalSimVerilatorIoTest.scala:210]: sure, was still debugging some stuff there, gonna get removed

File: core/src/main/scala/spinal/core/Enum.scala (hunk 3 above)
Comment (andreasWallner) [reply core/src/main/scala/spinal/core/Enum.scala:355]: thx

File: core/src/main/scala/spinal/core/Enum.scala (hunk 1 above)
Comment (andreasWallner) [reply core/src/main/scala/spinal/core/Enum.scala:361]: but no worries, easy to extend

File: core/src/main/scala/spinal/core/Misc.scala [hunk 5]
```diff
@@ -49,6 +49,26 @@ object log2Up {
...
+   * Encoding binary number in binary gray code
+   */
+  def encode(binary: BigInt): BigInt = binary ^ (binary >> 1)
+
+  /**
+   *  Decode binary gray encoded number to binary
+   */
+  def decode(gray: BigInt): BigInt = {
+    var binary = BigInt(0)
+    var bits = gray
```
Comment (numero-744) [inline core/src/main/scala/spinal/core/Misc.scala:63]: This is why I used FP style in my suggestion XD

File: core/src/main/scala/spinal/core/Misc.scala [hunk 6]
```diff
@@ -49,6 +49,26 @@ object log2Up {
   def apply(value : Int) : Int = apply(BigInt(value))
 }
 
+object Gray {
+  /**
+   * Encoding binary number in binary gray code
+   */
```
Comment (numero-744) [inline core/src/main/scala/spinal/core/Misc.scala:55]:
```suggestion
  /** Encoding binary number in binary gray code */
```

File: core/src/main/scala/spinal/core/Misc.scala [hunk 7]
```diff
@@ -49,6 +49,26 @@ object log2Up {
...
+  def encode(binary: BigInt): BigInt = binary ^ (binary >> 1)
+
+  /**
+   *  Decode binary gray encoded number to binary
+   */
```
Comment (numero-744) [inline core/src/main/scala/spinal/core/Misc.scala:60]:
```suggestion
  /** Decode binary gray encoded number to binary */
```

**Review by numero-744 — APPROVED** [review]
LGTM, just a few formatting suggestions

File: core/src/main/scala/spinal/core/Misc.scala (hunk 6 above)
Comment (andreasWallner) [reply core/src/main/scala/spinal/core/Misc.scala]: changed, but these are now the only two one-line comments formatted this way in this file

**Comment by kleinai** [comment]
Just my two cents but we already have gray coding functions in lib/utils. Should we do something with that rather than having duplicate code?

**Comment by numero-744** [comment]
@kleinai I think you are mentionning hardware functions, here it is elaboration-times (Int & BigInt instead of Bits)

**Comment by kleinai** [comment]
> @kleinai I think you are mentioning hardware functions, here it is elaboration-time (`Int` & `BigInt` instead of `Bits`)

Oops, I see that now, never mind.

**Comment by andreasWallner** [comment]
@Readon anything else that comes to mind, otherwise it would be ready from my side to merge

**Comment by Readon** [comment]
> @Readon anything else that comes to mind, otherwise it would be ready from my side to merge

To me, it's nice.
Good work.

**Review by Readon — APPROVED** [review]

**Pull request merged**